tools/upscaling_coefficients/upscaler_demo.py

```diff
@@ -0,0 +1,793 @@
+    0.022877571178541396, 0.452206427023829, 0.17567618137781177,
+    -0.04246845133905895], [-0.009019734953997752, -0.039000354140279844,
+    -0.040021010867420236, -0.03838632575427138, -0.02066407496750483]],
+    [[-0.0007159165805851705, -0.006389875713059274, -0.012319067611648212,
+    -0.010202069931803576, -0.0036010963945262565], [-0.007755474877032563,
+    -0.03942402104403912, -0.02118180471068666, -0.0278857406191104,
+    -0.029204772772557758], [-0.015459264122748742, -7.890574314417695e-05,
+    0.4909386897413151, 0.2868949551896562, -0.04562755195174026],
+    [-0.01450239495172347, -0.0057304550220541465, 0.4111930051936302,
+    0.23903219477825297, -0.04314593981787025], [-0.0064664894229144,
+    -0.03663285147762567, -0.03901507994141055, -0.037526010691823306,
+    -0.025194061504750526]], [[-0.0004507227386051219, -0.005422291349995998,
+    -0.01155272937910007, -0.01081748428801323, -0.00447631814814665],
+    [-0.005422291349995998, -0.03680917952171095, -0.022489385288542607,
+    -0.024475522375569655, -0.03324558280295302], [-0.011552729379100074,
+    -0.02248938528854261, 0.4272005024152728, 0.3575770812164143,
+    -0.0378381168526885], [-0.010817484288013228, -0.02447552237556967,
+    0.3575770812164143, 0.29895328454745285, -0.03706352644207268],
+    [-0.00447631814814665, -0.03324558280295302, -0.0378381168526885,
+    -0.03706352644207269, -0.029286133281073243]]]]
+}
+
+
+def convolution(pixels, coeffs, kernel_size):
+  pad_width = kernel_size//2
+  is_multi_channel = len(pixels.shape) > 2
+  padded_pixels = np.pad(
+      pixels, 2*[2*[pad_width]] + int(is_multi_channel)*[[0, 0]], mode='edge')
+  x, y,  = pixels.shape[:2]
+  convoluted = np.block([[np.einsum('rc...,RCrc->...RC',
+                                    padded_pixels[i-pad_width: i + pad_width + 1,
+                                                  j - pad_width: j + pad_width + 1],
+                                    coeffs)
+                          for j in range(pad_width, pad_width + y)]
+                         for i in range(pad_width, pad_width + x)])
+  if is_multi_channel:
+    return np.moveaxis(convoluted, 0, -1)
+  else:
+    return convoluted
+
+
+def main():
+    parser = argparse.ArgumentParser(
+        description="Upscaling of an image by a factor of 2, 4 or 8.")
+    parser.add_argument(
+        "--upscaling_factor",
+        type=int,
+        help="where N must be  2, 4 (default) or 8.",
+        nargs=1,
+        default=[4],
+        metavar='N')
+
+    parser.add_argument(
+        "input_filename",
+        type=str,
+        help="of the PNG image to be upscaled."
+    )
+    parser.add_argument(
+        "output_filename",
+        type=str,
+        help="where the upscaled image is written as PNG."
+    )
+
+    args = parser.parse_args()
+    upscaling_factor = args.upscaling_factor[0]
+    kernel_size = 5
+    assert upscaling_factor in [2, 4, 8]
```

Comment:
This will raise an AssertionError on user-provided-wrong-value, which looks weird to the user. Better:

if upscaling_factor not in (2, 4, 8):
  raise ValueError(...)

Reply:
Done